sequencing-server/src/lib/codegen/CommandEDSLPreface.ts

```diff
@@ -304,8 +434,91 @@ export class Sequence implements SeqJson {
+           * this needs to be inlined instead of extracted to a helper type alias so that monaco won't hide the underlying type behind the alias
+           */
+          locals: { [Index in Locals[number] as Index['name']]: Index['type'] };
+          parameters: { [Index in Parameters[number] as Index['name']]: Index['type'] };
+          // @ts-ignore : 'Step' found in JSON Spec
+        }) => Step[]);
+        // @ts-ignore : 'ImmediateCommand' found in JSON Spec
+        immediate_commands?: ImmediateCommand[];
+        // @ts-ignore : 'HardwareCommand' found in JSON Spec
+        hardware_commands?: HardwareCommand[];
+        // @ts-ignore : 'Request' found in JSON Spec
+        requests?: Request[];
+      }
+          // @ts-ignore : 'SeqJson' found in JSON Spec
+          | SeqJson,
+  ): Sequence {
+    if ('id' in opts) {
+      // @ts-ignore : 'SeqJson' found in JSON Spec
+      return new Sequence(opts as SeqJson);
+    } else {
+      /** Forcing the correct type for the undefined case */
+      const seqId = 'id' in opts ? opts.id : opts.seqId;
+      const metadata = opts.metadata;
+      const immediate_commands = opts.immediate_commands;
+      const hardware_commands = opts.hardware_commands;
+      const requests = opts.requests;
+
+      const locals = opts.locals ?? ([] as unknown as Locals);
+      const parameters = opts.parameters ?? ([] as unknown as Parameters);
+
+      // @ts-ignore
+      const localsMap = locals.reduce((acc, declaration) => {
+        /* ts ignore here because we're very much NOT doing what we're telling TS we are, but it makes for good UX
+         * we're not actually passing the type property of the declaration, but the whole declaration
+         * This ensures we get a readable type in the steps function
+         */
+        // @ts-ignore
+        const variable = Variable.new({ name: declaration.name, type: declaration.type });
+        variable.setKind('locals');
+        acc[declaration.name] = variable;
+        return acc;
+      }, {} as { [Index in Locals[number] as Index['name']]: Index['type'] });
+
+      // @ts-ignore
+      const parametersMap = parameters.reduce((acc, declaration) => {
+        // ts ignore here because we're very much NOT doing what we're telling TS we are, but it makes for good UX
+        // @ts-ignore
+        const variable = Variable.new({ name: declaration.name, type: declaration.type });
+        variable.setKind('parameters');
+        acc[declaration.name] = variable;
+        return acc;
+      }, {} as { [Index in Parameters[number] as Index['name']]: Index['type'] });
+
+      const steps =
+          typeof opts.steps === 'function' ? opts.steps({ locals: localsMap, parameters: parametersMap }) : opts.steps;
+
+      return new Sequence({
+        seqId,
+        metadata,
+        locals: locals.length !== 0 ? [locals[0], ...locals.slice(1)] : undefined,
```

Comment:
Interesting that we make these undefined instead of empty arrays.

Reply:
Yeah, this was to play nice with the SeqJson spec interface as an `empty []` isn't valid.

https://github.com/NASA-AMMOS/seq-json-schema/blob/develop/types.ts#L92

Reply:
I am going to have to keep the original implementation as the suggested change is not working with the json spec interface.

Reply:
Yeah, I wondered if it was going to affect the type checking. My guess is that the form we have specifies it as a non empty array since we essentially say the first element must exist.